When workers die, send rexi_EXIT message back to coordinating process

nickva opened the pull request:
Otherwise a coordinating process which uses an infinity timeout (in
particular couch_replicator changes feed process) will never notice a worker
died and will be stuck waiting forever.

In the case of the couch_replicator_manager this would manafest as it not
noticing new replication documents added to the replicator database and
never 'triggering' them.

FB-61505

FB-60382

davisp commented:
Subject line is too long. I suggest something like "Send rexi_EXIT message on stream timeout"

You should also add a note that the reason we exit normal is that in most cases its the coordinator that dies and we exit normal to avoid spamming logs with timeout errors since its usually a 24x and usually unimportant error since the real issue is why the coordinator disappeared.

Format for the FB ids is BugzId: 61505

Also you'll want to open a branch and PR against apache rexi so we don't lose this.

nickva committed:
Send rexi_EXIT message on stream timeout

Otherwise a coordinating process which uses an infinity timeout (in
particular couch_replicator changes feed process) will never notice a worker
died and will be stuck waiting forever.

This happened with a long polling, infinity timeout change feed
used by couch_replicator_manger -- callback processing took too long,
rexi worker timed out, but the main change feed process never noticed,
and got stuck waiting forever for more changes to arrive.

Note: exit(normal) is used instead of a timeout exit to avoid spamming logs.
Usually the coordinating process dies for some unrelated reason, so dumping
24x timeouts errors from worker proceess in the log would not be helpful.

BugzID: 61505

davisp commented:
+1
